plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+            throws IOException
+    {
+        GetObjectRequest.Builder getObjectRequestBuilder = GetObjectRequest.builder()
+                .bucket(getBucketName(file))
+                .key(keyFromUri(file));
+        S3RequestUtil.configureEncryption(secretKey, getObjectRequestBuilder);
+
+        try {
+            return new InputStreamSliceInput(s3Client.getObject(getObjectRequestBuilder.build(), ResponseTransformer.toInputStream()));
+        }
+        catch (AwsServiceException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public ExchangeStorageWriter createExchangeStorageWriter(URI file, Optional<SecretKey> secretKey)
+    {
+        String bucketName = getBucketName(file);
+        String key = keyFromUri(file);
+
+        return new S3ExchangeStorageWriter(s3Client, s3AsyncClient, bucketName, key, multiUploadPartSize, secretKey);
+    }
+
+    @Override
+    public boolean exists(URI file)
+            throws IOException
+    {
+        // Only used for commit marker files and doesn't need secretKey
+        return getHeadObjectResponse(file, Optional.empty()) != null;
+    }
+
+    @Override
+    public void createEmptyFile(URI file)
+            throws IOException
+    {
+        PutObjectRequest request = PutObjectRequest.builder()
+                .bucket(getBucketName(file))
+                .key(keyFromUri(file))
+                .build();
+
+        try {
+            s3Client.putObject(request, RequestBody.empty());
+        }
+        catch (AwsServiceException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public void deleteRecursively(URI uri)
+    {
+        deleteExecutor.submit(() -> deleteRecursivelySync(uri));
+    }
+
+    @Override
+    public Stream<URI> listFiles(URI dir)
+    {
+        List<S3Object> objects = listS3Directory(dir).contents();
+        return objects.stream().filter(object -> !object.key().endsWith(PATH_SEPARATOR)).map(object -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + object.key(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public Stream<URI> listDirectories(URI dir)
+    {
+        List<CommonPrefix> prefixes = listS3Directory(dir).commonPrefixes();
+        return prefixes.stream().map(prefix -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + prefix.prefix(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public long size(URI uri, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        checkArgument(!isDirectory(uri), "expected a file URI but got a directory URI");
```

Comment:
What if we get an URI which points to directory but without trailing `/`? 
Maybe we can skip this checkArgument and verify if we actually looked at the file based on the `HeadObjectResponse`. Is it encode there? Maybe we can get this info via `HeadObjectResponse.getMetadata()`?

Reply:
I intentionally did this to avoid calling S3 to check if a URI points to a directory or an object.

> What if we get an URI which points to directory but without trailing /?

Implementation should guarantee that all directories end with `/` but files don't

Reply:
Extra call just for a sanity check is  surely a no-go. I just wondered if we can get this information from `HeadObjectResponse` you do anyway.